### PR TITLE
Support bulkhead annotation on spring

### PR DESCRIPTION
**Pull request by lespinsideg**
- added Bulkhead annotation and BulkheadAspect on spring. Default aspect order is Retry -> Bulkhead -> RateLimiter -> CircuitBreaker.
- added auto configuration and events endpoints for spring-boot and spring-boot 2.  

Please review the code.

### CODE REVIEW

**Comment by RobWin** [comment]
PR looks good to me.
Could you please rebase the latest changes.

**Comment by RobWin** [comment]
I would love to add this feature to release 0.14.0

**Comment by lespinsideg** [comment]
@RobWin I resolved the conflict.

File: resilience4j-spring-boot/src/main/java/io/github/resilience4j/bulkhead/autoconfigure/BulkheadAutoConfiguration.java [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+@ConditionalOnClass(Bulkhead.class)
+@EnableConfigurationProperties(BulkheadProperties.class)
+@Import(BulkheadConfiguration.class)
+public class BulkheadAutoConfiguration {
+
+    private final BulkheadProperties bulkheadProperties;
+    private final BulkheadRegistry bulkheadRegistry;
+    private final ConfigurableBeanFactory beanFactory;
+
+    public BulkheadAutoConfiguration(BulkheadProperties circuitBreakerProperties, BulkheadRegistry bulkheadRegistry, ConfigurableBeanFactory beanFactory) {
```
Comment (RobWin) [inline resilience4j-spring-boot/src/main/java/io/github/resilience4j/bulkhead/autoconfigure/BulkheadAutoConfiguration.java]: Could you please rename it to bulkheadProperties

File: resilience4j-spring-boot2/src/main/java/io/github/resilience4j/bulkhead/autoconfigure/BulkheadAutoConfiguration.java [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+@ConditionalOnClass(Bulkhead.class)
+@EnableConfigurationProperties(BulkheadProperties.class)
+@Import(BulkheadConfiguration.class)
+public class BulkheadAutoConfiguration {
+
+    private final BulkheadProperties bulkheadProperties;
+    private final BulkheadRegistry bulkheadRegistry;
+    private final ConfigurableBeanFactory beanFactory;
+
+    public BulkheadAutoConfiguration(BulkheadProperties circuitBreakerProperties, BulkheadRegistry bulkheadRegistry, ConfigurableBeanFactory beanFactory) {
```
Comment (RobWin) [inline resilience4j-spring-boot2/src/main/java/io/github/resilience4j/bulkhead/autoconfigure/BulkheadAutoConfiguration.java]: Could you please rename it to bulkheadProperties

File: resilience4j-spring-boot/src/main/java/io/github/resilience4j/bulkhead/autoconfigure/BulkheadAutoConfiguration.java [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+import javax.annotation.PostConstruct;
+
+/**
+ * {@link org.springframework.boot.autoconfigure.EnableAutoConfiguration
+ * Auto-configuration} for resilience4j-bulkhead.
+ */
+@Configuration
+@ConditionalOnClass(Bulkhead.class)
+@EnableConfigurationProperties(BulkheadProperties.class)
+@Import(BulkheadConfiguration.class)
```
Comment (RobWin) [inline resilience4j-spring-boot/src/main/java/io/github/resilience4j/bulkhead/autoconfigure/BulkheadAutoConfiguration.java]: Could you please add `@AutoConfigureBefore(EndpointAutoConfiguration.class)`

File: resilience4j-spring-boot2/src/main/java/io/github/resilience4j/bulkhead/autoconfigure/BulkheadAutoConfiguration.java [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+
+/**
+ * {@link org.springframework.boot.autoconfigure.EnableAutoConfiguration
+ * Auto-configuration} for resilience4j-bulkhead.
+ */
+@Configuration
+@ConditionalOnClass(Bulkhead.class)
+@EnableConfigurationProperties(BulkheadProperties.class)
+@Import(BulkheadConfiguration.class)
+public class BulkheadAutoConfiguration {
```
Comment (RobWin) [inline resilience4j-spring-boot2/src/main/java/io/github/resilience4j/bulkhead/autoconfigure/BulkheadAutoConfiguration.java]: Could you please add `@AutoConfigureBefore(EndpointAutoConfiguration.class)`

File: resilience4j-spring-boot/src/main/java/io/github/resilience4j/bulkhead/monitoring/health/BulkheadHealthIndicator.java [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+import io.github.resilience4j.bulkhead.BulkheadConfig;
+import org.springframework.boot.actuate.health.Health;
+import org.springframework.boot.actuate.health.HealthIndicator;
+
+import java.util.Optional;
+
+/**
+ * A Spring Boot health indicators which adds the state of a Bulkhead and it's metrics to the health endpoints
+ */
+public class BulkheadHealthIndicator implements HealthIndicator {
```
Comment (RobWin) [inline resilience4j-spring-boot/src/main/java/io/github/resilience4j/bulkhead/monitoring/health/BulkheadHealthIndicator.java]: Do you think it makes sense to add a BulkheadHealthIndicator if the health state is always up? 
Metrics are already shown in the metrics endpoint.

Comment (lespinsideg) [reply resilience4j-spring-boot/src/main/java/io/github/resilience4j/bulkhead/monitoring/health/BulkheadHealthIndicator.java]: I removed `BulkheadHealthIndicator`.  Do you think it is better to show it as down state when a bulkhead is full?

**Comment by RobWin** [comment]
A test is failing. Did you rebase all changes?

**Comment by lespinsideg** [comment]
> A test is failing. Did you rebase all changes?

Yes, I did. But still some tests were failing. So I fixed them by separating DummyService.

**Pull request merged**